parser/src/aspects/redirection_parser.rs

```diff
@@ -0,0 +1,267 @@
+    use crate::ast::callable::{Call, Redir, Redirected, RedirFd, RedirOp};
+    use crate::ast::Expr;
+    use crate::ast::group::Block;
+    use crate::ast::literal::{Literal};
+    use crate::parse;
+    use crate::parser::Parser;
+
+    #[test]
+    fn expr_redirection() {
+        let tokens = lex("{ls; cd;} > /tmp/out");
+        let parsed = parse(tokens).expect("Failed to parse");
+        assert_eq!(
+            parsed,
+            vec![
+                Expr::Redirected(Redirected {
+                    expr: Box::new(Expr::Block(Block {
+                        expressions: vec![
+                            Expr::Call(Call {
+                                arguments: vec![
+                                    Expr::Literal("ls".into())
+                                ]
+                            }),
+                            Expr::Call(Call {
+                                arguments: vec![
+                                    Expr::Literal("cd".into())
+                                ]
+                            }),
+                        ]
+                    })),
+                    redirections: vec![
+                        Redir {
+                            operator: RedirOp::Write,
+                            fd: RedirFd::Default,
+                            operand: Expr::Literal("/tmp/out".into()),
+                        }
+                    ],
+                }),
+            ]
+        );
+    }
+
+    #[test]
+    fn call_redirection() {
+        let tokens = lex("ls> /tmp/out");
+        let parsed = Parser::new(tokens).call().expect("Failed to parse");
+        assert_eq!(
+            parsed,
+            Expr::Redirected(Redirected {
+                expr: Box::new(Expr::Call(Call {
+                    arguments: vec![Expr::Literal("ls".into())]
+                })),
+                redirections: vec![Redir {
+                    fd: RedirFd::Default,
+                    operator: RedirOp::Write,
+                    operand: Expr::Literal("/tmp/out".into()),
+                }],
+            })
+        );
+    }
+
+
+
+    #[test]
+    fn dupe_fd() {
+        let tokens = lex("ls>&2");
+        let parsed = Parser::new(tokens).call().expect("Failed to parse");
+        assert_eq!(
+            parsed,
+            Expr::Redirected(Redirected {
+                expr: Box::new(Expr::Call(Call {
+                    arguments: vec![Expr::Literal("ls".into())]
+                })),
+                redirections: vec![Redir {
+                    fd: RedirFd::Default,
+                    operator: RedirOp::FdOut,
+                    operand: Expr::Literal(Literal {
+                        lexme: "2",
+                        parsed: 2.into(),
+                    }),
+                }],
+            })
+        );
+    }
+
+
+    #[test]
+    fn escaped_call() {
+        let tokens = lex("grep -E regex \\; echo test");
+        let parsed = parse(tokens).expect("parsing error");
```

Comment:
Why is this in the redirection parser test? There is no redirection here.

Reply:
its a fail